### ISSUE TITLE
EntityHook: Duplicate EntityChanges per transaction

### ISSUE THREAD

**Issue by mpe85** [JetBrains/Exposed#723]
I have a question about the entity hook and the registered entity changes.

https://github.com/JetBrains/Exposed/blob/7e3c74232406b1d60609a5b8a856a7bffe4a8135/exposed-dao/src/main/kotlin/org/jetbrains/exposed/dao/EntityHook.kt#L37-L41

When a change gets registered there is a check if that change was already registered. But the code above only checks if the last added change is equal to the current change. If there was a different change in the meantime the duplicate change will be registered anyhow.
I'm not sure if this is on purpose but I would consider all changes on the same entity in the same transaction to be 'one' change in any case (also if there are changes on other entities in between).
So I would change the code to this:

```kotlin
fun registerChange(transaction: Transaction, change: EntityChange) { 
     if (change !in transaction.entityEvents) { 
         transaction.entityEvents.add(change) 
     } 
 } 
```

What is your opinion about this?

**Comment by Tapac** [comment]
The main idea is to keep all the events in the order they happened.

For example after that code execution:
```kotlin
val entity1 = FooEntity(1) 
entity1.delete()
val entity1copy = FooEntity(1)
```

There should be 2 `Created` event with the same id and 1 `Removed` event in between.

In that case, you could track what's happening with an entity within a transaction.

But at the moment `EntityHook` mechanism is not perfect (for example it doesn't track that was changed) and could be improved in the future.

**Comment by mpe85** [comment]
@Tapac Thanks for the feedback.
What would also help is to provide some kind of transaction identifier in the entity change event so the subscribers know which events were triggered within the same transaction. Maybe each transaction could have a unique UUID which is used for this. What so you think?

**Comment by mpe85** [comment]
@Tapac Mind if I provide a PR for it in the next few days?

**Comment by Tapac** [comment]
@mpe85 , sure! Will wait for a PR.

**Comment by mpe85** [comment]
@Tapac I created a PR: https://github.com/JetBrains/Exposed/pull/734
I made the `entityEvents` list to a private property of `Transaction`, and `EntityEvent` objects will now be craeated inside `Transaction.registerChange`. This way it is impossible to register an event with a different (wrong) transaction ID. It is better encapsulated this way in my opinion.

**Comment by Tapac** [comment]
Released in 0.20.1
